cpp/include/cudf/detail/utilities/hash_functions.cuh

```diff
@@ -365,18 +365,18 @@ struct SparkMurmurHash3_32 {
     return h;
   }
 
-  result_type __device__ inline operator()(Key const& key) const { return compute(key); }
+  [[nodiscard]] __device__ inline uint32_t getblock32(std::byte const* data,
+                                                      cudf::size_type offset) const
+  {
+    // Read a 4-byte value from the data pointer as individual bytes for safe
+    // unaligned access (very likely for string types).
+    auto block = reinterpret_cast<uint8_t const*>(data + offset);
+    return block[0] | (block[1] << 8) | (block[2] << 16) | (block[3] << 24);
```

Comment:
Future PR, but this would be more efficient to round up to the nearest 4B aligned address and do a 4B load and then shift the bits as appropriate to get them back in the expected order.

Reply:
You'd need to do _two_ 4B loads and shift both results together if you were reading 4B aligned addresses, right? This is meant to handle the case where the 4 bytes are not aligned (could cross 4B alignment boundaries).

Reply:
Yep, you'd have to do two when it's not already aligned.

Reply:
We have code for this sitting around in various places.

Reply:
Ok great. I'll figure that out in a future PR -- there are still other changes I have in mind, but this PR is needed to unblock other work so I'm trying to limit its scope.